### ISSUE TITLE
Handle SEP-10 authenticated transfers with TransferProvider

### ISSUE THREAD

**Issue by mikeyrf** [stellar/js-stellar-wallets#72]
SEP-6 now gives anchors the option to require SEP-10 based [client authentication](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0006.md#authentication) prior to requesting deposit/withdraw info.

We should likely add:

- An `AuthenticationProvider` to handle [SEP-10](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0010.md) challenge/response auth
- `authentication_required: boolean` fields to the [`RawInfoResponse`, `DepositAssetInfo`, `WithdrawAssetInfo`](https://github.com/stellar/js-stellar-wallets/blob/master/src/types/transfers.d.ts#L17) interfaces for each asset in deposit and withdrawal to reflect the new SEP-6 [`/info`](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0006.md#response-2) response fields
- An `AuthenticationResponse` interface with associated `type: TransferResponseType.authentication_required` to handle the `/deposit` and `/withdraw` auth required responses

Associated changes to the [`TransferProvider`](https://github.com/stellar/js-stellar-wallets/blob/master/src/transfers/TransferProvider.ts#L17) implementation will likely be needed to handle things properly.

**Comment by morleyzhi** [comment]
@mikeyrf Good point! Do you happen to know an anchor that supports SEP-10 auth, which we can use to test this on? (If not, we can try to find one or work with one to implement it, but if you have one in mind it would help me shortcut a few cycles.)

**Comment by mikeyrf** [comment]
[Us](https://authentication-dot-jewel-api-dev.appspot.com/?account=GCKM4UPHDEDDFRAR3DWQTUYIFDXB5JG4ARVI56RFEKUEBLTTPRT6AXHA) ;)

**Comment by mikeyrf** [comment]
@morleyzhi getting to implementing `/info`, etc. endpoints for SEP-6 as well, but that's likely in the next two weeks. I can get something demo-esque up and running likely in a few days

**Comment by morleyzhi** [comment]
Haha that's what I was just going to ask about. Let us know when those endpoints are ready. I or someone else might be able to start on this ticket in a week or two depending on how our roadmap lines up.

**Comment by mikeyrf** [comment]
Hey, [demo transfer server](https://transfer-dot-jewel-api-dev.appspot.com/info) is up to test against.

`/deposit` and `/info` endpoints should work fine. Since we're requiring authentication, the `account` query param on `/deposit` isn't needed (it's parsed from the JWT).

Some notes:
   - The transfer server expects a header for `auth_required` endpoints: `Authorization: Bearer <jwt>`
   - For `/deposit` requests, we're requiring query params `asset_code` and `type` (e.g. SWIFT)
   - To demo the edge cases, I've hard coded [this](https://github.com/dltjewel/go-jewel/blob/master/api/registration/types.go#L33) account to be KYC "registered". Any other auth token will throw a `403` err response with `interactive_customer_info_needed`
   - Authenticate in SEP-10 style against our [auth server](https://authentication-dot-jewel-api-dev.appspot.com/)

Most of our Go code for SEP-6 will be [open source](https://github.com/dltjewel/go-jewel/tree/master/api), if there's something you see that's weird (docs aren't all up to date tho :/)

**Comment by morleyzhi** [comment]
@mikeyrf I'm finally able to work on this this week!

I noticed that the demo account appears to be on Testnet, but the EURB / USDB withdrawal tokens' issuer is on mainnet (GAVEGF7WSL7MBMKUISFYEMC22AI7QVDTMNATEAYOR2BYUY3QM37IEWL2). Will deposits / withdrawals work on testnet? If so, is there a different testnet issuer address?

**Comment by morleyzhi** [comment]
@mikeyrf I'm trying to make a deposit with this call:

https://transfer-dot-jewel-api-dev.appspot.com/deposit?account=GCKM4UPHDEDDFRAR3DWQTUYIFDXB5JG4ARVI56RFEKUEBLTTPRT6AXHA&amount=15&asset_code=EUR&type=SWIFT

And the Authorization header set to "Bearer reallygreattoken", but I get back just the word "Unauthorized". Firstly, it'd be great if it could be a JSON object instead of a raw string, and secondly, do I have to use a specific bearer token to authenticate with the above user? As far as I can tell, it's the test user key you linked to.

**Comment by mikeyrf** [comment]
> @mikeyrf I'm finally able to work on this this week!

Yay!

> I noticed that the demo account appears to be on Testnet, but the EURB / USDB withdrawal tokens' issuer is on mainnet (GAVEGF7WSL7MBMKUISFYEMC22AI7QVDTMNATEAYOR2BYUY3QM37IEWL2). Will deposits / withdrawals work on testnet? If so, is there a different testnet issuer address?

It's all fake data at this point. None of the accounts or assets have been initialized or issued on either network.

We'll have instances on both testnet and mainnet, we're just still building all the public facing API stuff out on dev.

> Also, will the deposits and withdrawals automatically fill for demo purposes?

Not sure what you mean by fill. As in deposits sent via SWIFT actually going through? (don't do that yet pls)

**Comment by mikeyrf** [comment]
> @mikeyrf I'm trying to make a deposit with this call:
> 
> https://transfer-dot-jewel-api-dev.appspot.com/deposit?account=GCKM4UPHDEDDFRAR3DWQTUYIFDXB5JG4ARVI56RFEKUEBLTTPRT6AXHA&amount=15&asset_code=EUR&type=SWIFT
> 
> And the Authorization header set to "Bearer reallygreattoken", but I get back just the word "Unauthorized". Firstly, it'd be great if it could be a JSON object instead of a raw string,

@morleyzhi Agree on the JSON front. We had to fork [`go-chi/jwtauth`](https://github.com/dltjewel/jwtauth) exactly for this reason. There's just some small annoying issue with Google App Engine and go vendoring of forked repos I haven't figured out yet (it works locally).

Once that issue is resolved, the error response JSON (and similarly for other errors) will look like
```json
{
   "status": "Unauthorized",
   "error": "Authorization header invalid"
}
```

> and secondly, do I have to use a specific bearer token to authenticate with the above user? As far as I can tell, it's the test user key you linked to.

No, you don't need a specific bearer. But you need to use the SEP-10 authentication flow via our [auth server](https://authentication-dot-jewel-api-dev.appspot.com/?account=GCKM4UPHDEDDFRAR3DWQTUYIFDXB5JG4ARVI56RFEKUEBLTTPRT6AXHA): as in `GET` request with query param `account=...` to auth server to obtain the challenge, then sign the challenge with seed, and `POST` signed challenge to same endpoint.

If the auth flow hasn't been implemented yet for `js-stellar-wallets`, I can just feed you an example bearer token to include in auth headers while testing (they expire after one day)

**Comment by morleyzhi** [comment]
Ok, I understand about the token now. I'll work on implementing that in the Wallet SDK.

About filling: I meant that it'd be helpful to test deposits (and eventually withdrawals) end-to-end, where depositing actually adds tokens to my testnet account, without needing to do real bank transactions.

**Comment by mikeyrf** [comment]
Ah got it. End-to-end testing won't be ready for a few weeks.

Do you have access to a test environment for SWIFT? (I'd assume Dwolla for ACH)

**Comment by morleyzhi** [comment]
Can you authorize this testnet account?

GBF6SSPHXMVDRGF7OLM7LISI4FAFEU646FZRQOSB7OOCV4QZPILMHK53

That way I'll be able to sign challenge transactions. Thanks!

**Comment by mikeyrf** [comment]
[Account registered](https://github.com/dltjewel/go-jewel/blob/master/api/registration/types.go#L34) .. So you shouldn't get the customer registration redirect.

Or did you want a JWT for that account?

**Comment by morleyzhi** [comment]
Thanks! I'm planning to start on the authprovider right now, so no JWT needed.

**Comment by mikeyrf** [comment]
@morleyzhi I had to build out most of the spec for our own version of the `AuthProvider` to authenticate during our [onboarding flow](https://register-devnet.dltjewel.com/).

Would it help if I gave you access to that repo? Was going to port over the `trezor` related code to `js-stellar-wallets` eventually anyway

**Comment by dolcalmi** [comment]
> @mikeyrf Good point! Do you happen to know an anchor that supports SEP-10 auth, which we can use to test this on? (If not, we can try to find one or work with one to implement it, but if you have one in mind it would help me shortcut a few cycles.)

https://stellarport.io

**Comment by morleyzhi** [comment]
@dolcalmi Awesome, thanks!

Two questions: 
- I hit https://a3s-sandbox.api.stellarport.io/v2/GC6OWX3B4NSVUNKHHR6NDCBTFF7IPF6PPVCHXCD5TDTFWSB3LKB7QY55/info, but it looks like neither deposits nor withdraws have `authentication_required` set to true, is there a different asset issuer id I should hit to test SEP-10?
- If I request a deposit in the sandbox, does the deposit happen on my account automatically (without me sending real funds)? Or do I have to do something to induce the funds to appear?

Thanks for your help!

**Comment by istrau2** [comment]
@morleyzhi 

Hey @dolcalmi let me know about this thread.

- Sorry, looks like you are right. I've never seen the `authentication_required` flag in SEP-6, I guess it is relatively new. We will add the flag to our transfer server now.

- You need to actually make a deposit using TXRP. You can get yourself some Testnet Ripple at the [testnet ripple faucet](https://xrpl.org/xrp-test-net-faucet.html) and manage your testnet wallet from [this handy ripple wallet](https://ripplerm.github.io/ripple-wallet/) (just make sure you select the testnet network). Withdrawals also execute a withdrawal on the Ripple testnet. We chose Ripple because of the speed. Testing the deposit or withdrawal workflow can be done in a minute or two.

**Comment by morleyzhi** [comment]
Thanks @istrau2!

**Comment by morleyzhi** [comment]
@mikeyrf SEP-10 support is now in the latest version of @stellar/wallet-sdk, though it's untested live.

I noticed that your auth server returns a `tx` property:
https://authentication-dot-jewel-api-dev.appspot.com/?account=GCKM4UPHDEDDFRAR3DWQTUYIFDXB5JG4ARVI56RFEKUEBLTTPRT6AXHA

But the SEP (and the wallet SDK) expects the property name to be `transaction`:
https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0010.md#response

**Comment by morleyzhi** [comment]
Fixed in 142ca3432226cea514a97e2b1ab1ca22966875a9

**Comment by dolcalmi** [comment]
@morleyzhi are you validating the transaction? https://github.com/stellar/js-stellar-wallets/blob/master/src/KeyManager.ts#L237
